shared/sbl.js

```diff
@@ -0,0 +1,107 @@
+'use strict'
+
+type TypeFilter = (domain: string, selector: string, data: Array<*>) => ?boolean
+type TypeDomain = {
+  selectors: {[string]: Function},
+  domainFilters: Array<TypeFilter>,
+  selectorFilters: {[string]: TypeFilter},
+  data: Object
+}
+
+var domains: {[string]: TypeDomain} = {}
+var globalFilters: Array<TypeFilter> = []
+var pendingCalls = {} // allows sbl to be called regardless of module load order
+
+function sbl (path: string, ...data?: Array<*>) {
+  var [, name, selector] = /^([^/]+)(.+)/.exec(path)
+  var domain = domains[name]
+  if (!domain) {
+    if (!pendingCalls[name]) pendingCalls[name] = []
+    pendingCalls[name].push([path, data])
+    return
+  }
+  for (let gf of globalFilters) {
+    if (gf(name, selector, data) === false) return
+  }
+  for (let df of domain.domainFilters) {
+    if (df(name, selector, data) === false) return
+  }
+  var sf = domain.selectorFilters[selector]
+  if (sf && sf(name, selector, data) === false) return
+  return domain.selectors[selector].call(domain, ...data)
+}
+
+sbl.registerDomain = function (domain: string, selectors: {[string]: Function}) {
+  if (domains[domain]) throw new Error(`${domain} exists`)
+  domains[domain] = {selectors, data: {}, domainFilters: [], selectorFilters: {}}
+  // run any pendingCalls for this domain
+  if (pendingCalls[domain]) {
+    for (var [path, data] of pendingCalls[domain]) {
+      sbl(path, ...data)
+    }
+    delete pendingCalls[domain] // see sbl.throwIfPending
+  }
+}
+
+sbl.addDomainFilter = function (domain: string, filter: TypeFilter) {
+  domains[domain].domainFilters.push(filter)
+}
+
+sbl.setSelectorFilter = function (domain: string, sel: string, filter: TypeFilter) {
+  if (domains[domain].selectorFilters[sel]) throw new Error(`${domain}/${sel} filter exists`)
+  domains[domain].selectorFilters[sel] = filter
+}
+
+sbl.addGlobalFilter = function (filter: TypeFilter) {
+  globalFilters.push(filter)
+}
+
+// call at end of sbl setup, right before main run loop of app
+sbl.throwIfPending = function () {
+  var pending = Object.keys(pendingCalls)
+  if (pending.length > 0) {
+    throw new Error(`pending: ${JSON.stringify(pending)}`)
+  }
+}
+
+export default sbl
+
+// =======================
+// Common selectors APIs
+// TODO: move to a separate file
+// =======================
+
+function addListener (event, fn, type: 'listeners' | 'listenOnce') {
+  if (!this.data.events) {
+    this.data.events = {}
+  }
+  if (!this.data.events[event]) {
+    this.data.events[event] = {listeners: [], listenOnce: []}
+  }
+  this.data.events[event][type].push(fn)
+}
+
+const COMMON = {
```

Comment:
I this should be called `mixins`. If we call it 'common' we are giving it suggest meaning of common purpose to the application as whole and that might be a barrier to adding new ones to someone coming behind the initial developer. (even if it is only in nomenclature) Further, I think `EVENTS` should be called 'PUBSUB' because we've inadvertantly aliased this idea and functionality. `sbl.registerDomain('pubsub', sbl.COMMON.V1.EVENTS)` could read as `sbl.registerDomain('pubsub', sbl.MIXINS.V1.PUBSUB)` which could confirm that this mixin is the same as the pubsub domain/idea/functionality.